ci: Migrate CI scripts to CMake

hebasto opened the pull request:
For the same changes in the master branch with CI logs, please see https://github.com/bitcoin/bitcoin/pull/29790.

Still no Android-related changes.

fanquake commented on ci/test/00_setup_env_win64.sh:
```diff
@@ -16,4 +16,5 @@ export GOAL="deploy"
 # cross-compiling for Windows. https://sourceforge.net/p/mingw-w64/bugs/306/
 # https://github.com/mingw-w64/mingw-w64/commit/1690994f515910a31b9fb7c7bd3a52d4ba987abe
-export BITCOIN_CONFIG="--enable-reduce-exports --disable-gui-tests CXXFLAGS=-Wno-return-type"
+export BITCOIN_CONFIG="-DREDUCE_EXPORTS=ON -DBUILD_GUI_TESTS=OFF -DCMAKE_CXX_COMPILER=-Wno-return-type \
```
This is setting the compiler to be a warning flag, but I don't see any CI failures, so I guess this job isn't being run?

hebasto replied:
> This is setting the compiler to be a warning flag,

Thanks! This typo will be fixed in the next push.

> I don't see any CI failures, so I guess this job isn't being run?

For CI logs, please see [bitcoin#29790](https://github.com/bitcoin/bitcoin/pull/29790) as stated in the PR description.

fanquake commented on ci/test/00_setup_env_native_valgrind.sh:
```diff
@@ -14,4 +14,6 @@ export NO_DEPENDS=1
-export BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --with-gui=no CC=clang CXX=clang++ CFLAGS=-gdwarf-4 CXXFLAGS=-gdwarf-4"  # TODO enable GUI
+export BITCOIN_CONFIG="-DWITH_ZMQ=ON -DWARN_INCOMPATIBLE_BDB=OFF -DWITH_GUI=OFF \
+-DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ \
+-DCMAKE_C_FLAGS=-gdwarf-4 -DCMAKE_CXX_FLAGS=-gdwarf-4"  # TODO enable GUI
```
Looking at this job, and native_tsan. Is there a reason CXXFLAGS are sometimes going into  the "env"`CXXFLAGS`, and sometimes into `DCMAKE_CXX_FLAGS` ?

hebasto replied:
The former prepends generic non-configuration-specific flags being passed to the compiler. The latter overrides them.

fanquake replied:
Right, but we do we need to preprend (what are actually cppflags) here. I don't see why they can't go in cxxflags?

hebasto replied:
This diff mirrors the behavior from the master branch where the `CXXFLAGS` variable being set disables warning flags.

hebasto replied:
Reworked. It is consistent with the `ci/test/00_setup_env_native_fuzz_with_valgrind.sh` now.

fanquake commented on ci/test/00_setup_env_native_previous_releases.sh:
```diff
@@ -16,5 +16,6 @@ export RUN_UNIT_TESTS_SEQUENTIAL="true"
-CFLAGS=\"-g0 -O2 -funsigned-char\" CPPFLAGS='-DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE' CXXFLAGS=\"-g0 -O2 -funsigned-char\""
+export BITCOIN_CONFIG="-DWITH_ZMQ=ON -DWITH_GUI=Qt5 -DREDUCE_EXPORTS=ON -DCMAKE_BUILD_TYPE=Debug \
+-DCMAKE_C_FLAGS='-funsigned-char' -DCMAKE_CXX_FLAGS='-DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE -funsigned-char' \
+-DCMAKE_C_FLAGS_DEBUG='-g0 -O2' -DCMAKE_CXX_FLAGS_DEBUG='-g0 -O2'"
```
Why do we need to set `*_FLAGS` and `*_FLAGS_DEBUG` separately here?

hebasto replied:
> Why do we need to set `*_FLAGS` and `*_FLAGS_DEBUG` separately here?

The configuration-specific variants `*_FLAGS_DEBUG` are proper CMake variables to handle optimization and debugging flags.

fanquake replied:
Right, but if we are using the debug mode, why can't what is being put into `CMAKE_C{XX}_FLAGS` just also be put into `CMAKE_C{XX}_FLAGS_DEBUG`.

hebasto replied:
Setting the `CMAKE_CXX_FLAGS` mirrors behaviour in the master branch where the `CXXFLAGS` variable being set disables warning flags.

hebasto replied:
> Right, but if we are using the debug mode, why can't what is being put into `CMAKE_C{XX}_FLAGS` just also be put into `CMAKE_C{XX}_FLAGS_DEBUG`.

Thanks! Done.

fanquake commented on ci/test/00_setup_env_native_tidy.sh:
```diff
@@ -9,12 +9,14 @@ export LC_ALL=C.UTF-8
-export BITCOIN_CONFIG="CC=clang-${TIDY_LLVM_V} CXX=clang++-${TIDY_LLVM_V} --with-incompatible-bdb --disable-hardening CFLAGS='-O0 -g0' CXXFLAGS='-O0 -g0 -I/usr/lib/llvm-${TIDY_LLVM_V}/lib/clang/${TIDY_LLVM_V}/include'"
+export BITCOIN_CONFIG="-DWARN_INCOMPATIBLE_BDB=OFF -DHARDENING=OFF \
+-DCMAKE_C_COMPILER=clang-${TIDY_LLVM_V} -DCMAKE_CXX_COMPILER=clang++-${TIDY_LLVM_V} \
+-DCMAKE_C_FLAGS_RELWITHDEBINFO='-O0 -g0' -DCMAKE_CXX_FLAGS_RELWITHDEBINFO='-O0 -g0'"
```
Why is this job using `CMAKE_C_FLAGS_RELWITHDEBINFO`?

hebasto replied:
There is `CFLAGS='-O0 -g0'` in the master branch. In CMake, optimization and debugging flags are handled by configuration-specific flags. This job uses the default configuration:
```
Build type (configuration):
 - CMAKE_BUILD_TYPE ................... RelWithDebInfo
```

That's why `-DCMAKE_C_FLAGS_RELWITHDEBINFO='-O0 -g0'` is a CMake's analogue of Autotools's `CFLAGS='-O0 -g0'`.

fanquake reviewed (commented):
Had a quick look over.

hebasto commented:
https://github.com/hebasto/bitcoin/pull/142#discussion_r1554393305 has been addressed.

See Cirrus CI logs in https://github.com/bitcoin/bitcoin/pull/29790 -- https://cirrus-ci.com/build/4887272447803392.

hebasto commented:
Rebased.

hebasto commented on ci/test/wrap-wine.sh:
```diff
@@ -6,7 +6,7 @@
 export LC_ALL=C.UTF-8
 
-for b_name in {"${BASE_OUTDIR}/bin"/*,src/secp256k1/*tests,src/minisketch/test{,-verify},src/univalue/{test_json,unitester,object}}.exe; do
+for b_name in {"${BASE_OUTDIR}/bin"/*,src/minisketch/test{,-verify},src/univalue/{test_json,unitester,object}}.exe; do
```
In CMake build system, we agreed to skip tests for subtree code.

hebasto replied:
Thanks! Adjusted.

hebasto commented on ci/test/03_test_script.sh:
```diff
@@ -99,17 +99,20 @@ if [ "$DOWNLOAD_PREVIOUS_RELEASES" = "true" ]; then
 ccache --zero-stats
 PRINT_CCACHE_STATISTICS="ccache --version | head -n 1 && ccache --show-stats"
 
+mkdir -p "${BASE_BUILD_DIR}"
```
Thanks! Done.

An excerpt from https://cirrus-ci.com/task/5556699505885184:
>     -- Build files have been written to: /ci_container_base/ci/scratch/build-i686-pc-linux-gnu

hebasto commented:
Rebased.

The last commit is https://github.com/bitcoin/bitcoin/commit/c06b2bb9d61df469190452b39c3aaa88069aeaa9 cherry-picked from the https://github.com/bitcoin/bitcoin/pull/29790. So reviewers can inspect CI logs there.

It follows the following [approach](https://github.com/hebasto/bitcoin/pull/184#issuecomment-2117308393):
> ... we agreed on using `APPEND_*FLAGS` variables as the last resort only, without advertising them to users and attempting to use native CMake's variables as wide as possible...

---

@maflcko Do you mind having another look at this PR please?

theuni commented on ci/test/00_setup_env_native_previous_releases.sh:
```diff
@@ -16,5 +16,6 @@ export RUN_UNIT_TESTS_SEQUENTIAL="true"
-export BITCOIN_CONFIG="--enable-zmq --with-gui=qt5 --enable-reduce-exports --enable-debug \
-CFLAGS=\"-g0 -O2 -funsigned-char\" CPPFLAGS='-DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE' CXXFLAGS=\"-g0 -O2 -funsigned-char\""
+export BITCOIN_CONFIG="-DWITH_ZMQ=ON -DBUILD_GUI=ON -DREDUCE_EXPORTS=ON -DCMAKE_BUILD_TYPE=Debug \
+-DCMAKE_C_FLAGS_DEBUG='-g0 -O2 -funsigned-char' \
```
Shouldn't the `funsigned-char` here be `-DCMAKE_C[XX]_FLAGS` since it's not part of the optims?

hebasto replied:
In https://github.com/hebasto/bitcoin/pull/142#discussion_r1554395773, @fanquake expressed the opposite opinion :) 

The result will be the same for both approaches. The one you ask about is just a bit more verbose.

Do you want me to apply it?

theuni replied:
Hah.

I suppose it's up to @fanquake. Imo using both vars is more correct, but I don't feel too strongly.

fanquake replied:
> I suppose it's up to @fanquake. Imo using both vars is more correct, but I don't feel too strongly.

Now that we've (re-)changed approach, we can split them back out into the all the different vars.

hebasto replied:
Thanks! done in the recent push.

hebasto commented:
@fanquake requested offline to:
> ... take the compiler flags out of the CC and CXX defines.

Done. This PR and https://github.com/bitcoin/bitcoin/pull/29790 have been updated simultaneously.

hebasto commented:
Addressed @theuni's https://github.com/hebasto/bitcoin/pull/142#discussion_r1605534766.

This PR and [bitcoin#29790](https://github.com/bitcoin/bitcoin/pull/29790) have been updated simultaneously.

TheCharlatan commented on ci/test/00_setup_env_native_fuzz_with_msan.sh:
```diff
@@ -18,7 +18,7 @@ export PACKAGES="ninja-build"
 export GOAL="install"
 # _FORTIFY_SOURCE is not compatible with MSAN.
-export BITCOIN_CONFIG="--enable-fuzz --with-sanitizers=fuzzer,memory CPPFLAGS='-DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE -U_FORTIFY_SOURCE'"
+export BITCOIN_CONFIG="-DFUZZ=ON -DSANITIZERS=fuzzer,memory -DAPPEND_CPPFLAGS='-DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE -U_FORTIFY_SOURCE'"
```
Are we using the `APPEND_*FLAGS` here just because setting `-DCMAKE_CXX_FLAGS` would clobber the `CXXFLAGS`?

hebasto replied:
No.

`-DAPPEND_CPPFLAGS='-U_FORTIFY_SOURCE'` is required to override `-D_FORTIFY_SOURCE=3` from https://github.com/hebasto/bitcoin/blob/05784622feb15252a75dad64cdd5723e612e3f14/CMakeLists.txt#L505-L507

hebasto replied:
FWIW, there was https://github.com/bitcoin/bitcoin/pull/30140.

fanquake replied:
I guess the answer here is that given how hardening has been implemented in the CMake branch, it's impossible to override any of the flag using the proper CMake vars?

hebasto replied:
> I guess the answer here is that given how hardening has been implemented in the CMake branch, it's impossible to override any of the flags using the proper CMake vars?

That's correct.

fanquake replied:
That seems like a problem in it's own right, but I guess can be followed up with separately.

theuni replied:
Isn't this the exact use case we added the APPEND flags for, though?

fanquake replied:
We'll probably just need to document which flags happen to only work with those flags in some way, because at the moment, it's buried in CMake implementation detail.

theuni replied:
On second glance, `-DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE` is going into `DCMAKE_CXX_FLAGS` in most other places. If nothing else, we should be consistent with where we stick vars to avoid confusion.

hebasto replied:
> On second glance, `-DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE` is going into `DCMAKE_CXX_FLAGS` in most other places. If nothing else, we should be consistent with where we stick vars to avoid confusion.

Done.

TheCharlatan commented on ci/test/00_setup_env_native_previous_releases.sh:
```diff
@@ -16,5 +16,6 @@ export RUN_UNIT_TESTS_SEQUENTIAL="true"
-export BITCOIN_CONFIG="--enable-zmq --with-gui=qt5 --enable-reduce-exports --enable-debug \
-CFLAGS=\"-g0 -O2 -funsigned-char\" CPPFLAGS='-DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE' CXXFLAGS=\"-g0 -O2 -funsigned-char\""
+export BITCOIN_CONFIG="-DWITH_ZMQ=ON -DBUILD_GUI=ON -DREDUCE_EXPORTS=ON -DCMAKE_BUILD_TYPE=Debug \
+-DCMAKE_C_FLAGS='-funsigned-char' -DCMAKE_C_FLAGS_DEBUG='-g0 -O2' \
```
Why is this setting both the generic and build-type specific flags?

hebasto replied:
There were related discussions [here](https://github.com/hebasto/bitcoin/pull/142#discussion_r1554395773) and [here](https://github.com/hebasto/bitcoin/pull/142#discussion_r1605534766).

TheCharlatan commented on cmake/leveldb.cmake:
```diff
@@ -99,3 +99,7 @@ target_link_libraries(leveldb PRIVATE
 )
+
+set_target_properties(leveldb PROPERTIES
+  EXPORT_COMPILE_COMMANDS OFF
```
Why do we need to set this for some of our targets?

hebasto replied:
By default, when `EXPORT_COMPILE_COMMANDS` is `ON`, CMake generates a `compile_commands.json`, which is used in the "tidy" CI task. On the master branch, `compile_commands.json` is generated by the `bear` tool.

We do not want to analyze code from subtrees, so disabling `EXPORT_COMPILE_COMMANDS` in there.

hebasto replied:
Isn't it supposed to be done in upstream repositories?

hebasto commented on ci/test/00_setup_env_native_asan.sh:
```diff
@@ -17,10 +17,11 @@ else
 
 export CONTAINER_NAME=ci_native_asan
-export PACKAGES="systemtap-sdt-dev clang-18 llvm-18 libclang-rt-18-dev python3-zmq qtbase5-dev qttools5-dev-tools libevent-dev libboost-dev libdb5.3++-dev libminiupnpc-dev libnatpmp-dev libzmq3-dev libqrencode-dev libsqlite3-dev ${BPFCC_PACKAGE}"
+export PACKAGES="systemtap-sdt-dev clang-18 llvm-18 libclang-rt-18-dev python3-zmq qtbase5-dev qttools5-dev qttools5-dev-tools libevent-dev libboost-dev libdb5.3++-dev libminiupnpc-dev libnatpmp-dev libzmq3-dev libqrencode-dev libsqlite3-dev ${BPFCC_PACKAGE}"
```
It follows the build docs: https://github.com/hebasto/bitcoin/blob/ecd23656db174adef61d3bd753d02698c3528192/doc/build-unix.md?plain=1#L84

Otherwise, CMakes fails to find `Qt5LinguistTools`.

TheCharlatan replied:
Right, please resolve.

TheCharlatan commented:
Changes seem sane to me, but given this changes much of the CI I might be missing something.

hebasto commented:
I'm going to merge this PR shortly.

theuni commented on ci/test/00_setup_env_i686_multiprocess.sh:
```diff
@@ -13,6 +13,6 @@ export PACKAGES="llvm clang g++-multilib"
-export BITCOIN_CONFIG="--enable-debug CC='clang -m32' CXX='clang++ -m32' \
-CPPFLAGS='-DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE' CXXFLAGS='-Wno-error=documentation'"
+export BITCOIN_CONFIG="-DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER='clang;-m32' -DCMAKE_CXX_COMPILER='clang++;-m32' \
+-DCMAKE_CXX_FLAGS='-DBOOST_MULTI_INDEX_ENABLE_SAFE_MODE -Wno-error=documentation'"
```
Are these not going into APPEND flags because we don't add `-Wdocumentation` ourselves? Meaning that the primary use-case for APPEND_CXXFLAGS is specifically when we:
- Add a flag in our own buildsystem
- Want that flag turned off by the user at build-time

?

If so, it would be nice to document exactly that.

theuni replied:
Wait, we do add `-Wdocumentation` though. So why doesn't this have to be in APPEND_CXXFLAGS?

hebasto replied:
We have to use `APPEND_CXXFLAGS` for `-Wno-documentation`. 

`-Wno-error=documentation` works with `CMAKE_CXX_FLAGS` just fine.

theuni replied:
Ah, ok. I would've thought that order matters there. But that makes sense.

theuni commented:
Agree with what @TheCharlatan said. I left a few nits, but this is a pretty huge change and I don't feel that I can review the whole thing and compare it to master with much confidence. I'm ok merging it (after addressing comments) and dealing with any fallout in a follow-up.

hebasto commented:
> I'm ok merging it (after addressing comments) and dealing with any fallout in a follow-up.

https://github.com/bitcoin/bitcoin/pull/29790 will continue to run the CI for the `cmake-staging` branch. So we'll be able to observe any overlooked problem.
